src/main/kotlin/com/safehill/kclient/tasks/syncing/ThreadInteractionSync.kt

```diff
@@ -0,0 +1,150 @@
+            threadId = threadId,
+            conversationThreadAssetDtos = assets
+        )
+    }
+
+    private suspend fun TextMessage.notifyTextMessage() {
+        when (anchorType) {
+            InteractionAnchor.THREAD -> {
+                threadInteractionSyncListener.didReceiveTextMessages(
+                    messageDtos = listOf(this.toMessageDTO()),
+                    threadId = this.anchorId
+                ).also {
+                    serverProxy.localServer.insertMessages(
+                        messages = listOf(this.toMessageDTO()),
+                        threadId = this.anchorId
+                    )
+                }
+            }
+
+            InteractionAnchor.GROUP -> {}
+        }
+    }
+
+    private fun TextMessage.toMessageDTO() = MessageOutputDTO(
+        interactionId = interactionId,
+        senderUserIdentifier = senderPublicIdentifier,
+        inReplyToAssetGlobalIdentifier = inReplyToAssetGlobalIdentifier,
+        encryptedMessage = encryptedMessage,
+        createdAt = sentAt,
+        inReplyToInteractionId = inReplyToInteractionId
+    )
+
+    private suspend fun syncThreadInteractionSummary() {
+        coroutineScope {
+            val interactionsSummary = runCatchingPreservingCancellationException {
+                serverProxy.topLevelInteractionsSummary()
+            }
+            interactionsSummary.onSuccess { interactionsSummaryDTO ->
+                val allThreads = interactionsSummaryDTO.summaryByThreadId.map {
+                    it.value.thread
+                }
+                val localThreads = serverProxy.localServer.listThreads()
+                deleteUnwantedThreadsLocally(
+                    allThreads = allThreads,
+                    localThreads = localThreads
+                )
+                launch {
+                    threadInteractionSyncListener.didFetchRemoteThreadSummary(
+                        interactionsSummaryDTO.summaryByThreadId
+                    )
+                }
+            }
+        }
+    }
+
+    private fun CoroutineScope.deleteUnwantedThreadsLocally(
+        allThreads: List<ConversationThreadOutputDTO>,
+        localThreads: List<ConversationThreadOutputDTO>
+    ) {
+        launch {
+            val threadIdsToRemoveLocally = localThreads.filter { localThread ->
+                localThread.threadId !in allThreads.map { it.threadId }
+            }.map { it.threadId }
+            if (threadIdsToRemoveLocally.isNotEmpty()) {
+                serverProxy.localServer.deleteThreads(threadIdsToRemoveLocally)
+            }
+        }
+    }
+}
+
+interface InteractionSyncListener {
+    suspend fun didUpdateThreadsList(threadDTOs: List<ConversationThreadOutputDTO>)
+
+    suspend fun didAddThread(threadDTO: ConversationThreadOutputDTO)
+
+    suspend fun didReceiveTextMessages(messageDtos: List<MessageOutputDTO>, threadId: String)
+
+    suspend fun didReceivePhotoMessages(
```

Comment:
How do you distinguish between photo messages and assets shared with the people in the thread?

Reply:
As of now, only the photo messages are notified by this method and  whereas assets shared outside of thread is notified to us by the RemoteDownloadOperation.